File: tests/e2e/java/src/test/java/dev/feast/E2ETest.java
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2018-2020 The Feast Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.feast;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.assertEquals;

import com.gojek.feast.FeastClient;
import com.gojek.feast.Row;
import feast.common.auth.credentials.GoogleAuthCredentials;
import feast.proto.types.ValueProto.Value;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * E2E tests Java SDK retrieval with Feast. Expects Feast to have a registered Feature Set
 * 'customer_transactions' that is already ingested with data and ready for retrieval. See
 * tests/e2e/python/redis/basic-ingest-redis-serving.py for test setup code. Properties used to
 * configure test params: feast.serving.host - hostname of the feast serving instance to connect to.
 * feast.serving.port - port of the feast serving instance to connect to. feast.auth.enabled - if
 * true, uses authentication to connect to Feast serving.
 */
public class E2ETest {
  private static FeastClient client;
  private static boolean isAuthenticationEnabled;

  @BeforeAll
  public static void setup() throws IOException {
    // Read test params from properties.
    Properties properties = System.getProperties();
    String servingHost = properties.getProperty("feast.serving.host");
    int servingPort = Integer.parseInt(properties.getProperty("feast.serving.port"));
    isAuthenticationEnabled = Boolean.parseBoolean(properties.getProperty("feast.auth.enabled"));

    // Setup clients for test.
    if (!isAuthenticationEnabled) {
      client = FeastClient.create(servingHost, servingPort);
    } else {
      // Uses GoogleAuthCredentials to authenticate
      // Expects Google Application Credentials to configured with Service Account JSON.
      String audience = String.format("%s:%d", servingHost, servingPort);
      client =
          FeastClient.createAuthenticated(
              servingHost, servingPort, new GoogleAuthCredentials(Map.of("audience", audience)));
    }
  }

  @Test
  public void shouldGetOnlineFeatures() {
    List<String> featureRefs =
        List.of("daily_transactions", "customer_transactions:total_transactions");

    List<Row> featureRows =
        client.getOnlineFeatures(
            featureRefs,
            List.of(Row.create().set("customer_id", Value.newBuilder().setInt64Val(1).build())));
    assertThat(featureRows.size(), equalTo(1));

    // Check feature values and statuses retrieved are correct.
    // Values are ingested in tests/e2e/python/redis/basic-ingest-redis-serving.py.
    Map<String, Value> fields = featureRows.get(0).getFields();
    System.out.println(fields.keySet());
    featureRefs.stream()
        .map(ref -> fields.get(ref))
        .forEach(value -> assertEquals(value.getFloatVal(), 1.0f, 0.00001f));
  }
}

File: tests/e2e/go/e2e_test.go
package main

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"testing"

	"github.com/feast-dev/feast/sdk/go"
	"github.com/feast-dev/feast/sdk/go/protos/feast/serving"
	"github.com/feast-dev/feast/sdk/go/protos/feast/types"
	"github.com/google/go-cmp/cmp"
)

type TestParams struct {
	servingHost string
	servingPort int
	useAuth     bool
}

// Parse test params from environment variables
// SERVING_HOST - hostname of the Feast Serving instance to connect to.
// SERVING_PORT - port of Feast Serving instance to connect to.
// USE_AUTH - whether to use authentication when connect to Feast Serving.f
func parseParams() (*TestParams, error) {
	port, err := strconv.Atoi(os.Getenv("SERVING_PORT"))
	if err != nil {
		return nil, fmt.Errorf("Could not parse SERVING_PORT as a int")
	}
	useAuth, err := strconv.ParseBool(os.Getenv("USE_AUTH"))
	if err != nil {
		return nil, fmt.Errorf("Could not parse USE_AUTH as a boolean")
	}
	return &TestParams{
		servingHost: os.Getenv("SERVING_HOST"),
		servingPort: port,
		useAuth:     useAuth,
	}, nil
}

// Configure and return a Feast Client based on the given test params.
func configClient(params *TestParams) (*feast.GrpcClient, error) {
	if params.useAuth {
		cred, err := feast.NewGoogleCredential("localhost:6566")
		if err != nil {
			return nil, err
		}
		return feast.NewAuthGrpcClient(params.servingHost, params.servingPort, feast.SecurityConfig{
			Credential: cred,
		})
	} else {
		return feast.NewGrpcClient(params.servingHost, params.servingPort)
	}
}

// E2E Test Retrieval using Go SDK with Feast. Expects Feast to have a registered Feature Set
// 'customer_transactions' that is already ingested with data and ready for retrieval. See
// tests/e2e/python/redis/basic-ingest-redis-serving.py for test setup code.
func TestGetOnlineFeatures(t *testing.T) {
	// setup test client
	params, err := parseParams()
	if err != nil {
		t.Error(err)
	}
	cli, err := configClient(params)
	if err != nil {
		t.Error(err)
	}

	tt := []struct {
		name string
		req  *feast.OnlineFeaturesRequest
		want *feast.OnlineFeaturesResponse
	}{
		{
			name: "Valid Get OnlineFeaturesRequest",
			req: &feast.OnlineFeaturesRequest{
				Features: []string{
					"daily_transactions",
					"customer_transactions:total_transactions",
				},
				Entities: []feast.Row{
					{"customer_id": feast.Int64Val(1)},
				},
			},
			want: &feast.OnlineFeaturesResponse{
				RawResponse: &serving.GetOnlineFeaturesResponse{
					FieldValues: []*serving.GetOnlineFeaturesResponse_FieldValues{
						{
							Fields: map[string]*types.Value{
								"customer_id":        feast.Int64Val(1),
								"daily_transactions": feast.FloatVal(1.0),
								"customer_transactions:total_transactions": feast.FloatVal(1.0),
							},
							Statuses: map[string]serving.GetOnlineFeaturesResponse_FieldStatus{
								"customer_id":        serving.GetOnlineFeaturesResponse_PRESENT,
								"daily_transactions": serving.GetOnlineFeaturesResponse_PRESENT,
								"customer_transactions:total_transactions": serving.GetOnlineFeaturesResponse_PRESENT,
							},
						},
					},
				},
			},
		},
	}

	ctx := context.Background()
	for _, tc := range tt {
		resp, err := cli.GetOnlineFeatures(ctx, tc.req)
		if err != nil {
			t.Error(err)
		}

		if !cmp.Equal(resp.RawResponse.String(), tc.want.RawResponse.String()) {
			t.Errorf("got: \n%v\nwant:\n%v", resp.RawResponse.String(), tc.want.RawResponse.String())
		}
	}
}
